doc/reference/trd104-syscalls.md

```diff
@@ -235,57 +245,60 @@ The 6 classes are:
 | Memop            |        5         |
 | Exit             |        6         |
 
-All of the system call classes except Yield and Exit are non-blocking. When a userspace
-process calls a Subscribe, Command, Allow, Read-Only Allow, or Memop syscall,
-the kernel will not put it on a wait queue. Instead, it will return immediately
-upon completion. The kernel scheduler may decide to suspend the process due to
-a timeslice expiration or the kernel thread being runnable, but the system calls
-themselves do not block. If an operation is long-running (e.g., I/O), its completion
-is signaled by a callback (see the Subscribe call in 4.2).
+All of the system call classes except Yield and Exit are
+non-blocking. When a userspace process calls a Subscribe, Command,
+Allow, Read-Only Allow, or Memop syscall, the kernel will not put it
+on a wait queue. Instead, it will return immediately upon
+completion. The kernel scheduler may decide to suspend the process due
+to a timeslice expiration or the kernel thread being runnable, but the
+system calls themselves do not block. If an operation is long-running
+(e.g., I/O), its completion is signaled by an upcall (see the
+Subscribe call in 4.2).
```

Comment:
I would have written this that system calls themselves _do_ block. For example, every memop blocks the process from running until the entire memop operation has completed. Only then will the kernel reschedule the process.

Also, with the new scheduler trait, there is no guarantee that a process will not be put on a wait queue after calling a syscall.

Reply:
That's not traditionally what blocking means. Non-blocking I/O calls, for example, still return a return value. Usually blocking means that calling the function can cause the process to be descheduled and put on a wait queue, or forced to wait in some way.

Reply:
I agree with @phil-levis  that "blocking" implies blocking on I/O (or some other long operation).

@bradjc is right that there is no type- or implementation- guarantee, but this _is_ the semantics of system calls, and it _should_ be implemented by any scheduler (unless of course that scheduler is exploring ways in which different semantics might be better). In other words, it is the semantic that a Tock kernel promises to processes.